src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -963,6 +974,19 @@ ExecuteModifyTasks(List *taskList, bool expectResults, ParamListInfo paramListIn
 		return 0;
 	}
 
+	foreach(taskCell, taskList)
+	{
+		Task *task = (Task *) lfirst(taskCell);
+		ShardInterval *shardInterval = LoadShardInterval(task->anchorShardId);
+		Oid relationId = shardInterval->relationId;
+
+		/*
+		 * If we are dealing with partitioned table, we also need to lock its
+		 * partitions.
+		 */
+		LockPartitionRelationsIfPartitioned(relationId, RowExclusiveLock);
```

Comment:
But how can we know whether an element in the list belongs to a partitioned table without doing all that work: